server/apps/packages/package_service.py

```diff
@@ -206,6 +206,13 @@ def check_for_circular_reference(self, package, item_id):
             message = 'Trying to create a circular reference to: ' + item_id
             logger.error(message)
             raise ValidationError(message)
+        else:
+            # keep checking in the hierarchy
+            for d in package.get(LINKED_IN_PACKAGES, []):
+                if 'package' in d:
```

Comment:
To save a level of indentation, this condition can be embedded into the `for` loop expression (the next line is a bit long and will benefit from that).